gui/widgets/groupedlist.cpp

```diff
@@ -0,0 +1,427 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "common/system.h"
+#include "common/frac.h"
+#include "common/tokenizer.h"
+
+#include "gui/widgets/groupedlist.h"
+#include "gui/widgets/scrollbar.h"
+#include "gui/dialog.h"
+#include "gui/gui-manager.h"
+
+#include "gui/ThemeEval.h"
+
+#define kGroupTag -2
+#define isGroupHeader(x) (x <= kGroupTag)
+#define indexToGroupID(x) (kGroupTag - x)
+
+namespace GUI {
+
+GroupedListWidget::GroupedListWidget(Dialog *boss, const String &name, const Common::U32String &tooltip, uint32 cmd)
+	: ListWidget(boss, name, tooltip, cmd) {
+
+}
+
+void GroupedListWidget::setList(const U32StringArray &list, const ColorList *colors) {
+	if (_editMode && _caretVisible)
+		drawCaret(true);
+
+	// Copy everything
+	_dataList = list;
+	_list = list;
+
+	_filter.clear();
+	_listIndex.clear();
+	_listColors.clear();
+
+	if (colors) {
+		_listColors = *colors;
+		assert(_listColors.size() == _dataList.size());
+	}
+
+	int size = list.size();
+	if (_currentPos >= size)
+		_currentPos = size - 1;
+	if (_currentPos < 0)
+		_currentPos = 0;
+	_selectedItem = -1;
+	_editMode = false;
+	g_system->setFeatureState(OSystem::kFeatureVirtualKeyboard, false);
+	groupByAttribute();
+	scrollBarRecalc();
+}
+
+void GroupedListWidget::setAttributeValues(const U32StringArray &attrValues) {
+	_attributeValues = attrValues;
+	if (!attrValues.empty())
+		assert(_attributeValues.size() == _dataList.size());
+}
+
+void GroupedListWidget::append(const String &s, ThemeEngine::FontColor color) {
+	if (_dataList.size() == _listColors.size()) {
+		// If the color list has the size of the data list, we append the color.
+		_listColors.push_back(color);
+	} else if (!_listColors.size() && color != ThemeEngine::kFontColorNormal) {
```

Comment:
"_listColors.size() == 0" is clearer IMHO, as we aren't checking a boolean, but an actual value. May be a matter of taste though. This could perhaps be written in terms of empty() instead?

Reply:
I have replaced it with empty() in 0e8aa5b, seems much more readable.